{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "98b3cabc-60aa-491f-a04a-c04365e25f4d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eeab478e-e9c7-4cda-8d0e-15a6d432991d",
   "metadata": {},
   "source": [
    "# Exercise 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3aa3a60-1d25-4cd9-8747-f611ce112e4e",
   "metadata": {},
   "source": [
    "Read in data stored in the file \"10_pandas_idxstats.tsv\" and save to variable called `idx_df`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "id": "50d7e891-4b15-4d68-b877-f341e08b08fb",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MT</td>\n",
       "      <td>16571</td>\n",
       "      <td>226</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>249250621</td>\n",
       "      <td>1202908</td>\n",
       "      <td>11586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>243199373</td>\n",
       "      <td>1179409</td>\n",
       "      <td>11950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>198022430</td>\n",
       "      <td>956939</td>\n",
       "      <td>7894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>191154276</td>\n",
       "      <td>841065</td>\n",
       "      <td>8496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>180915260</td>\n",
       "      <td>839376</td>\n",
       "      <td>7313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>171115067</td>\n",
       "      <td>851357</td>\n",
       "      <td>6930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>159138663</td>\n",
       "      <td>639203</td>\n",
       "      <td>7728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>146364022</td>\n",
       "      <td>951606</td>\n",
       "      <td>8874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>141213431</td>\n",
       "      <td>603020</td>\n",
       "      <td>6937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>135534747</td>\n",
       "      <td>692429</td>\n",
       "      <td>10325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>135006516</td>\n",
       "      <td>660885</td>\n",
       "      <td>7635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>133851895</td>\n",
       "      <td>728753</td>\n",
       "      <td>6993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13</td>\n",
       "      <td>115169878</td>\n",
       "      <td>523741</td>\n",
       "      <td>4061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>14</td>\n",
       "      <td>107349540</td>\n",
       "      <td>480831</td>\n",
       "      <td>4255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>15</td>\n",
       "      <td>102531392</td>\n",
       "      <td>447954</td>\n",
       "      <td>3936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16</td>\n",
       "      <td>90354753</td>\n",
       "      <td>483073</td>\n",
       "      <td>5439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>17</td>\n",
       "      <td>81195210</td>\n",
       "      <td>532624</td>\n",
       "      <td>5339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>18</td>\n",
       "      <td>78077248</td>\n",
       "      <td>319957</td>\n",
       "      <td>4373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>19</td>\n",
       "      <td>59128983</td>\n",
       "      <td>407712</td>\n",
       "      <td>5882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>20</td>\n",
       "      <td>63025520</td>\n",
       "      <td>321007</td>\n",
       "      <td>3280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>21</td>\n",
       "      <td>48129895</td>\n",
       "      <td>175650</td>\n",
       "      <td>2351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>22</td>\n",
       "      <td>51304566</td>\n",
       "      <td>224761</td>\n",
       "      <td>2206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>X</td>\n",
       "      <td>155270560</td>\n",
       "      <td>490861</td>\n",
       "      <td>5788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Y</td>\n",
       "      <td>59373566</td>\n",
       "      <td>10375</td>\n",
       "      <td>3028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1174990</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     0          1        2        3\n",
       "0   MT      16571      226        2\n",
       "1    1  249250621  1202908    11586\n",
       "2    2  243199373  1179409    11950\n",
       "3    3  198022430   956939     7894\n",
       "4    4  191154276   841065     8496\n",
       "5    5  180915260   839376     7313\n",
       "6    6  171115067   851357     6930\n",
       "7    7  159138663   639203     7728\n",
       "8    8  146364022   951606     8874\n",
       "9    9  141213431   603020     6937\n",
       "10  10  135534747   692429    10325\n",
       "11  11  135006516   660885     7635\n",
       "12  12  133851895   728753     6993\n",
       "13  13  115169878   523741     4061\n",
       "14  14  107349540   480831     4255\n",
       "15  15  102531392   447954     3936\n",
       "16  16   90354753   483073     5439\n",
       "17  17   81195210   532624     5339\n",
       "18  18   78077248   319957     4373\n",
       "19  19   59128983   407712     5882\n",
       "20  20   63025520   321007     3280\n",
       "21  21   48129895   175650     2351\n",
       "22  22   51304566   224761     2206\n",
       "23   X  155270560   490861     5788\n",
       "24   Y   59373566    10375     3028\n",
       "25   *          0        0  1174990"
      ]
     },
     "execution_count": 202,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idx_df = pd.read_csv(\"idxstats.tsv\",sep='\\t',header=None)\n",
    "idx_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5b894e7-12ab-4f26-bd3c-6b841f4ec601",
   "metadata": {},
   "source": [
    "Set the column names to “name”, “length”, “mapped”, and “unmapped”."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "id": "e4c95c64-7383-41df-bd32-9ef60c399929",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>length</th>\n",
       "      <th>mapped</th>\n",
       "      <th>unmapped</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MT</td>\n",
       "      <td>16571</td>\n",
       "      <td>226</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>249250621</td>\n",
       "      <td>1202908</td>\n",
       "      <td>11586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>243199373</td>\n",
       "      <td>1179409</td>\n",
       "      <td>11950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>198022430</td>\n",
       "      <td>956939</td>\n",
       "      <td>7894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>191154276</td>\n",
       "      <td>841065</td>\n",
       "      <td>8496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>180915260</td>\n",
       "      <td>839376</td>\n",
       "      <td>7313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>171115067</td>\n",
       "      <td>851357</td>\n",
       "      <td>6930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>159138663</td>\n",
       "      <td>639203</td>\n",
       "      <td>7728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>146364022</td>\n",
       "      <td>951606</td>\n",
       "      <td>8874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>141213431</td>\n",
       "      <td>603020</td>\n",
       "      <td>6937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>135534747</td>\n",
       "      <td>692429</td>\n",
       "      <td>10325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>135006516</td>\n",
       "      <td>660885</td>\n",
       "      <td>7635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>133851895</td>\n",
       "      <td>728753</td>\n",
       "      <td>6993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13</td>\n",
       "      <td>115169878</td>\n",
       "      <td>523741</td>\n",
       "      <td>4061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>14</td>\n",
       "      <td>107349540</td>\n",
       "      <td>480831</td>\n",
       "      <td>4255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>15</td>\n",
       "      <td>102531392</td>\n",
       "      <td>447954</td>\n",
       "      <td>3936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16</td>\n",
       "      <td>90354753</td>\n",
       "      <td>483073</td>\n",
       "      <td>5439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>17</td>\n",
       "      <td>81195210</td>\n",
       "      <td>532624</td>\n",
       "      <td>5339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>18</td>\n",
       "      <td>78077248</td>\n",
       "      <td>319957</td>\n",
       "      <td>4373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>19</td>\n",
       "      <td>59128983</td>\n",
       "      <td>407712</td>\n",
       "      <td>5882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>20</td>\n",
       "      <td>63025520</td>\n",
       "      <td>321007</td>\n",
       "      <td>3280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>21</td>\n",
       "      <td>48129895</td>\n",
       "      <td>175650</td>\n",
       "      <td>2351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>22</td>\n",
       "      <td>51304566</td>\n",
       "      <td>224761</td>\n",
       "      <td>2206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>X</td>\n",
       "      <td>155270560</td>\n",
       "      <td>490861</td>\n",
       "      <td>5788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Y</td>\n",
       "      <td>59373566</td>\n",
       "      <td>10375</td>\n",
       "      <td>3028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1174990</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   name     length   mapped  unmapped\n",
       "0    MT      16571      226         2\n",
       "1     1  249250621  1202908     11586\n",
       "2     2  243199373  1179409     11950\n",
       "3     3  198022430   956939      7894\n",
       "4     4  191154276   841065      8496\n",
       "5     5  180915260   839376      7313\n",
       "6     6  171115067   851357      6930\n",
       "7     7  159138663   639203      7728\n",
       "8     8  146364022   951606      8874\n",
       "9     9  141213431   603020      6937\n",
       "10   10  135534747   692429     10325\n",
       "11   11  135006516   660885      7635\n",
       "12   12  133851895   728753      6993\n",
       "13   13  115169878   523741      4061\n",
       "14   14  107349540   480831      4255\n",
       "15   15  102531392   447954      3936\n",
       "16   16   90354753   483073      5439\n",
       "17   17   81195210   532624      5339\n",
       "18   18   78077248   319957      4373\n",
       "19   19   59128983   407712      5882\n",
       "20   20   63025520   321007      3280\n",
       "21   21   48129895   175650      2351\n",
       "22   22   51304566   224761      2206\n",
       "23    X  155270560   490861      5788\n",
       "24    Y   59373566    10375      3028\n",
       "25    *          0        0   1174990"
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns=[\"name\",\"length\", \"mapped\",\"unmapped\"]\n",
    "idx_df.columns=columns\n",
    "idx_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c0bef42-2054-44c5-8b46-a901ded62d69",
   "metadata": {},
   "source": [
    "Apply the `.describe()` and `.info()` methods to the `idx_df` object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "id": "f6f2e69e-529b-40fb-8ef5-a5c976ac879c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>length</th>\n",
       "      <th>mapped</th>\n",
       "      <th>unmapped</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>2.600000e+01</td>\n",
       "      <td>2.600000e+01</td>\n",
       "      <td>2.600000e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>1.190652e+08</td>\n",
       "      <td>5.602201e+05</td>\n",
       "      <td>5.106119e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>6.568174e+07</td>\n",
       "      <td>3.330361e+05</td>\n",
       "      <td>2.292555e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>6.678845e+07</td>\n",
       "      <td>3.426832e+05</td>\n",
       "      <td>4.109500e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>1.245109e+08</td>\n",
       "      <td>5.281825e+05</td>\n",
       "      <td>6.406000e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>1.581716e+08</td>\n",
       "      <td>8.117202e+05</td>\n",
       "      <td>7.852500e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>2.492506e+08</td>\n",
       "      <td>1.202908e+06</td>\n",
       "      <td>1.174990e+06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             length        mapped      unmapped\n",
       "count  2.600000e+01  2.600000e+01  2.600000e+01\n",
       "mean   1.190652e+08  5.602201e+05  5.106119e+04\n",
       "std    6.568174e+07  3.330361e+05  2.292555e+05\n",
       "min    0.000000e+00  0.000000e+00  2.000000e+00\n",
       "25%    6.678845e+07  3.426832e+05  4.109500e+03\n",
       "50%    1.245109e+08  5.281825e+05  6.406000e+03\n",
       "75%    1.581716e+08  8.117202e+05  7.852500e+03\n",
       "max    2.492506e+08  1.202908e+06  1.174990e+06"
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idx_df.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "id": "dfed8bde-c2fe-4a5d-884c-c0c198b57b03",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 26 entries, 0 to 25\n",
      "Data columns (total 4 columns):\n",
      " #   Column    Non-Null Count  Dtype \n",
      "---  ------    --------------  ----- \n",
      " 0   name      26 non-null     object\n",
      " 1   length    26 non-null     int64 \n",
      " 2   mapped    26 non-null     int64 \n",
      " 3   unmapped  26 non-null     int64 \n",
      "dtypes: int64(3), object(1)\n",
      "memory usage: 964.0+ bytes\n"
     ]
    }
   ],
   "source": [
    "idx_df.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5157b644-dd1b-4037-9c8c-cbde34903fd9",
   "metadata": {},
   "source": [
    "Look at the shape and index attributes of `idx_df`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11e61faa-d5a8-42c3-a2ca-942ddaa4323f",
   "metadata": {},
   "source": [
    "From `idx_df` select rows 6-12 and create a new variable called `idx_df2`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "id": "699df557-493d-41d6-9501-238d7423ac07",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>length</th>\n",
       "      <th>mapped</th>\n",
       "      <th>unmapped</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>180915260</td>\n",
       "      <td>839376</td>\n",
       "      <td>7313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>171115067</td>\n",
       "      <td>851357</td>\n",
       "      <td>6930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>159138663</td>\n",
       "      <td>639203</td>\n",
       "      <td>7728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>146364022</td>\n",
       "      <td>951606</td>\n",
       "      <td>8874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>141213431</td>\n",
       "      <td>603020</td>\n",
       "      <td>6937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>135534747</td>\n",
       "      <td>692429</td>\n",
       "      <td>10325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>135006516</td>\n",
       "      <td>660885</td>\n",
       "      <td>7635</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   name     length  mapped  unmapped\n",
       "5     5  180915260  839376      7313\n",
       "6     6  171115067  851357      6930\n",
       "7     7  159138663  639203      7728\n",
       "8     8  146364022  951606      8874\n",
       "9     9  141213431  603020      6937\n",
       "10   10  135534747  692429     10325\n",
       "11   11  135006516  660885      7635"
      ]
     },
     "execution_count": 206,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idx_df2 = idx_df.iloc[5:12,:]\n",
    "idx_df2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33443e42-286f-4a82-a1a1-64689640dcfb",
   "metadata": {},
   "source": [
    "From `idx_df` make a new dataframe `chr_mapped` consisting of columns 1 and 3 (“name” and ”mapped”). Save to a new file called `\"mapped_reads_per_chr.tsv\"` - Make sure the output is TAB separated."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "id": "c67c0ad8-5420-4987-9304-36ab84887d09",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   name   mapped\n",
      "0    MT      226\n",
      "1     1  1202908\n",
      "2     2  1179409\n",
      "3     3   956939\n",
      "4     4   841065\n",
      "5     5   839376\n",
      "6     6   851357\n",
      "7     7   639203\n",
      "8     8   951606\n",
      "9     9   603020\n",
      "10   10   692429\n",
      "11   11   660885\n",
      "12   12   728753\n",
      "13   13   523741\n",
      "14   14   480831\n",
      "15   15   447954\n",
      "16   16   483073\n",
      "17   17   532624\n",
      "18   18   319957\n",
      "19   19   407712\n",
      "20   20   321007\n",
      "21   21   175650\n",
      "22   22   224761\n",
      "23    X   490861\n",
      "24    Y    10375\n",
      "25    *        0\n"
     ]
    }
   ],
   "source": [
    "chr_mapped = idx_df.loc[:,[\"name\", \"mapped\"]]\n",
    "print(chr_mapped)\n",
    "chr_mapped.to_csv(\"mapped_reads_per_chr.tsv\", sep =\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24d6699b-5f98-40fe-aafc-87131d0288bd",
   "metadata": {},
   "source": [
    "# Exercise 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ea35385-0c04-4157-8b22-5e6cd40d0d00",
   "metadata": {},
   "source": [
    "Create a new column for `idx_df` called reads_per_base by dividing number of mapped reads by the `length` column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "id": "b60b08ed-ec1d-4a3d-bdbd-7370d8c47bd4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>length</th>\n",
       "      <th>mapped</th>\n",
       "      <th>unmapped</th>\n",
       "      <th>reads_per_base</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MT</td>\n",
       "      <td>16571</td>\n",
       "      <td>226</td>\n",
       "      <td>2</td>\n",
       "      <td>0.013638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>249250621</td>\n",
       "      <td>1202908</td>\n",
       "      <td>11586</td>\n",
       "      <td>0.004826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>243199373</td>\n",
       "      <td>1179409</td>\n",
       "      <td>11950</td>\n",
       "      <td>0.004850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>198022430</td>\n",
       "      <td>956939</td>\n",
       "      <td>7894</td>\n",
       "      <td>0.004832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>191154276</td>\n",
       "      <td>841065</td>\n",
       "      <td>8496</td>\n",
       "      <td>0.004400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>180915260</td>\n",
       "      <td>839376</td>\n",
       "      <td>7313</td>\n",
       "      <td>0.004640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>171115067</td>\n",
       "      <td>851357</td>\n",
       "      <td>6930</td>\n",
       "      <td>0.004975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>159138663</td>\n",
       "      <td>639203</td>\n",
       "      <td>7728</td>\n",
       "      <td>0.004017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>146364022</td>\n",
       "      <td>951606</td>\n",
       "      <td>8874</td>\n",
       "      <td>0.006502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>141213431</td>\n",
       "      <td>603020</td>\n",
       "      <td>6937</td>\n",
       "      <td>0.004270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>135534747</td>\n",
       "      <td>692429</td>\n",
       "      <td>10325</td>\n",
       "      <td>0.005109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>135006516</td>\n",
       "      <td>660885</td>\n",
       "      <td>7635</td>\n",
       "      <td>0.004895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>133851895</td>\n",
       "      <td>728753</td>\n",
       "      <td>6993</td>\n",
       "      <td>0.005444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13</td>\n",
       "      <td>115169878</td>\n",
       "      <td>523741</td>\n",
       "      <td>4061</td>\n",
       "      <td>0.004548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>14</td>\n",
       "      <td>107349540</td>\n",
       "      <td>480831</td>\n",
       "      <td>4255</td>\n",
       "      <td>0.004479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>15</td>\n",
       "      <td>102531392</td>\n",
       "      <td>447954</td>\n",
       "      <td>3936</td>\n",
       "      <td>0.004369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16</td>\n",
       "      <td>90354753</td>\n",
       "      <td>483073</td>\n",
       "      <td>5439</td>\n",
       "      <td>0.005346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>17</td>\n",
       "      <td>81195210</td>\n",
       "      <td>532624</td>\n",
       "      <td>5339</td>\n",
       "      <td>0.006560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>18</td>\n",
       "      <td>78077248</td>\n",
       "      <td>319957</td>\n",
       "      <td>4373</td>\n",
       "      <td>0.004098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>19</td>\n",
       "      <td>59128983</td>\n",
       "      <td>407712</td>\n",
       "      <td>5882</td>\n",
       "      <td>0.006895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>20</td>\n",
       "      <td>63025520</td>\n",
       "      <td>321007</td>\n",
       "      <td>3280</td>\n",
       "      <td>0.005093</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>21</td>\n",
       "      <td>48129895</td>\n",
       "      <td>175650</td>\n",
       "      <td>2351</td>\n",
       "      <td>0.003649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>22</td>\n",
       "      <td>51304566</td>\n",
       "      <td>224761</td>\n",
       "      <td>2206</td>\n",
       "      <td>0.004381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>X</td>\n",
       "      <td>155270560</td>\n",
       "      <td>490861</td>\n",
       "      <td>5788</td>\n",
       "      <td>0.003161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Y</td>\n",
       "      <td>59373566</td>\n",
       "      <td>10375</td>\n",
       "      <td>3028</td>\n",
       "      <td>0.000175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1174990</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   name     length   mapped  unmapped  reads_per_base\n",
       "0    MT      16571      226         2        0.013638\n",
       "1     1  249250621  1202908     11586        0.004826\n",
       "2     2  243199373  1179409     11950        0.004850\n",
       "3     3  198022430   956939      7894        0.004832\n",
       "4     4  191154276   841065      8496        0.004400\n",
       "5     5  180915260   839376      7313        0.004640\n",
       "6     6  171115067   851357      6930        0.004975\n",
       "7     7  159138663   639203      7728        0.004017\n",
       "8     8  146364022   951606      8874        0.006502\n",
       "9     9  141213431   603020      6937        0.004270\n",
       "10   10  135534747   692429     10325        0.005109\n",
       "11   11  135006516   660885      7635        0.004895\n",
       "12   12  133851895   728753      6993        0.005444\n",
       "13   13  115169878   523741      4061        0.004548\n",
       "14   14  107349540   480831      4255        0.004479\n",
       "15   15  102531392   447954      3936        0.004369\n",
       "16   16   90354753   483073      5439        0.005346\n",
       "17   17   81195210   532624      5339        0.006560\n",
       "18   18   78077248   319957      4373        0.004098\n",
       "19   19   59128983   407712      5882        0.006895\n",
       "20   20   63025520   321007      3280        0.005093\n",
       "21   21   48129895   175650      2351        0.003649\n",
       "22   22   51304566   224761      2206        0.004381\n",
       "23    X  155270560   490861      5788        0.003161\n",
       "24    Y   59373566    10375      3028        0.000175\n",
       "25    *          0        0   1174990             NaN"
      ]
     },
     "execution_count": 177,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idx_df[\"reads_per_base\"]=idx_df[\"mapped\"]/idx_df[\"length\"]\n",
    "idx_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "168b05a5-36c0-440b-a893-638002cc684d",
   "metadata": {},
   "source": [
    "Subset the rows that have `reads_per_base` values greater than 0.005"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "id": "c3db0a14-9fc2-4407-a276-a30a8d236902",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>length</th>\n",
       "      <th>mapped</th>\n",
       "      <th>unmapped</th>\n",
       "      <th>reads_per_base</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MT</td>\n",
       "      <td>16571</td>\n",
       "      <td>226</td>\n",
       "      <td>2</td>\n",
       "      <td>0.013638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>146364022</td>\n",
       "      <td>951606</td>\n",
       "      <td>8874</td>\n",
       "      <td>0.006502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>135534747</td>\n",
       "      <td>692429</td>\n",
       "      <td>10325</td>\n",
       "      <td>0.005109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>133851895</td>\n",
       "      <td>728753</td>\n",
       "      <td>6993</td>\n",
       "      <td>0.005444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16</td>\n",
       "      <td>90354753</td>\n",
       "      <td>483073</td>\n",
       "      <td>5439</td>\n",
       "      <td>0.005346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>17</td>\n",
       "      <td>81195210</td>\n",
       "      <td>532624</td>\n",
       "      <td>5339</td>\n",
       "      <td>0.006560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>19</td>\n",
       "      <td>59128983</td>\n",
       "      <td>407712</td>\n",
       "      <td>5882</td>\n",
       "      <td>0.006895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>20</td>\n",
       "      <td>63025520</td>\n",
       "      <td>321007</td>\n",
       "      <td>3280</td>\n",
       "      <td>0.005093</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   name     length  mapped  unmapped  reads_per_base\n",
       "0    MT      16571     226         2        0.013638\n",
       "8     8  146364022  951606      8874        0.006502\n",
       "10   10  135534747  692429     10325        0.005109\n",
       "12   12  133851895  728753      6993        0.005444\n",
       "16   16   90354753  483073      5439        0.005346\n",
       "17   17   81195210  532624      5339        0.006560\n",
       "19   19   59128983  407712      5882        0.006895\n",
       "20   20   63025520  321007      3280        0.005093"
      ]
     },
     "execution_count": 178,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idx_df[idx_df[\"reads_per_base\"] > 0.005]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24b27de6-4bb7-45f8-bf56-b3278ad00913",
   "metadata": {},
   "source": [
    "Create a new column called `length_Mbs` with length divided by 1 million (i.e., length in Megabases)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "id": "6710f1ca-6e6b-4235-b30b-9d2e501f1a37",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>length</th>\n",
       "      <th>mapped</th>\n",
       "      <th>unmapped</th>\n",
       "      <th>reads_per_base</th>\n",
       "      <th>length_Mbs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MT</td>\n",
       "      <td>16571</td>\n",
       "      <td>226</td>\n",
       "      <td>2</td>\n",
       "      <td>0.013638</td>\n",
       "      <td>0.016571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>249250621</td>\n",
       "      <td>1202908</td>\n",
       "      <td>11586</td>\n",
       "      <td>0.004826</td>\n",
       "      <td>249.250621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>243199373</td>\n",
       "      <td>1179409</td>\n",
       "      <td>11950</td>\n",
       "      <td>0.004850</td>\n",
       "      <td>243.199373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>198022430</td>\n",
       "      <td>956939</td>\n",
       "      <td>7894</td>\n",
       "      <td>0.004832</td>\n",
       "      <td>198.022430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>191154276</td>\n",
       "      <td>841065</td>\n",
       "      <td>8496</td>\n",
       "      <td>0.004400</td>\n",
       "      <td>191.154276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>180915260</td>\n",
       "      <td>839376</td>\n",
       "      <td>7313</td>\n",
       "      <td>0.004640</td>\n",
       "      <td>180.915260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>171115067</td>\n",
       "      <td>851357</td>\n",
       "      <td>6930</td>\n",
       "      <td>0.004975</td>\n",
       "      <td>171.115067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>159138663</td>\n",
       "      <td>639203</td>\n",
       "      <td>7728</td>\n",
       "      <td>0.004017</td>\n",
       "      <td>159.138663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>146364022</td>\n",
       "      <td>951606</td>\n",
       "      <td>8874</td>\n",
       "      <td>0.006502</td>\n",
       "      <td>146.364022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>141213431</td>\n",
       "      <td>603020</td>\n",
       "      <td>6937</td>\n",
       "      <td>0.004270</td>\n",
       "      <td>141.213431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>135534747</td>\n",
       "      <td>692429</td>\n",
       "      <td>10325</td>\n",
       "      <td>0.005109</td>\n",
       "      <td>135.534747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>135006516</td>\n",
       "      <td>660885</td>\n",
       "      <td>7635</td>\n",
       "      <td>0.004895</td>\n",
       "      <td>135.006516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>133851895</td>\n",
       "      <td>728753</td>\n",
       "      <td>6993</td>\n",
       "      <td>0.005444</td>\n",
       "      <td>133.851895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13</td>\n",
       "      <td>115169878</td>\n",
       "      <td>523741</td>\n",
       "      <td>4061</td>\n",
       "      <td>0.004548</td>\n",
       "      <td>115.169878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>14</td>\n",
       "      <td>107349540</td>\n",
       "      <td>480831</td>\n",
       "      <td>4255</td>\n",
       "      <td>0.004479</td>\n",
       "      <td>107.349540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>15</td>\n",
       "      <td>102531392</td>\n",
       "      <td>447954</td>\n",
       "      <td>3936</td>\n",
       "      <td>0.004369</td>\n",
       "      <td>102.531392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16</td>\n",
       "      <td>90354753</td>\n",
       "      <td>483073</td>\n",
       "      <td>5439</td>\n",
       "      <td>0.005346</td>\n",
       "      <td>90.354753</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>17</td>\n",
       "      <td>81195210</td>\n",
       "      <td>532624</td>\n",
       "      <td>5339</td>\n",
       "      <td>0.006560</td>\n",
       "      <td>81.195210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>18</td>\n",
       "      <td>78077248</td>\n",
       "      <td>319957</td>\n",
       "      <td>4373</td>\n",
       "      <td>0.004098</td>\n",
       "      <td>78.077248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>19</td>\n",
       "      <td>59128983</td>\n",
       "      <td>407712</td>\n",
       "      <td>5882</td>\n",
       "      <td>0.006895</td>\n",
       "      <td>59.128983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>20</td>\n",
       "      <td>63025520</td>\n",
       "      <td>321007</td>\n",
       "      <td>3280</td>\n",
       "      <td>0.005093</td>\n",
       "      <td>63.025520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>21</td>\n",
       "      <td>48129895</td>\n",
       "      <td>175650</td>\n",
       "      <td>2351</td>\n",
       "      <td>0.003649</td>\n",
       "      <td>48.129895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>22</td>\n",
       "      <td>51304566</td>\n",
       "      <td>224761</td>\n",
       "      <td>2206</td>\n",
       "      <td>0.004381</td>\n",
       "      <td>51.304566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>X</td>\n",
       "      <td>155270560</td>\n",
       "      <td>490861</td>\n",
       "      <td>5788</td>\n",
       "      <td>0.003161</td>\n",
       "      <td>155.270560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Y</td>\n",
       "      <td>59373566</td>\n",
       "      <td>10375</td>\n",
       "      <td>3028</td>\n",
       "      <td>0.000175</td>\n",
       "      <td>59.373566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1174990</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   name     length   mapped  unmapped  reads_per_base  length_Mbs\n",
       "0    MT      16571      226         2        0.013638    0.016571\n",
       "1     1  249250621  1202908     11586        0.004826  249.250621\n",
       "2     2  243199373  1179409     11950        0.004850  243.199373\n",
       "3     3  198022430   956939      7894        0.004832  198.022430\n",
       "4     4  191154276   841065      8496        0.004400  191.154276\n",
       "5     5  180915260   839376      7313        0.004640  180.915260\n",
       "6     6  171115067   851357      6930        0.004975  171.115067\n",
       "7     7  159138663   639203      7728        0.004017  159.138663\n",
       "8     8  146364022   951606      8874        0.006502  146.364022\n",
       "9     9  141213431   603020      6937        0.004270  141.213431\n",
       "10   10  135534747   692429     10325        0.005109  135.534747\n",
       "11   11  135006516   660885      7635        0.004895  135.006516\n",
       "12   12  133851895   728753      6993        0.005444  133.851895\n",
       "13   13  115169878   523741      4061        0.004548  115.169878\n",
       "14   14  107349540   480831      4255        0.004479  107.349540\n",
       "15   15  102531392   447954      3936        0.004369  102.531392\n",
       "16   16   90354753   483073      5439        0.005346   90.354753\n",
       "17   17   81195210   532624      5339        0.006560   81.195210\n",
       "18   18   78077248   319957      4373        0.004098   78.077248\n",
       "19   19   59128983   407712      5882        0.006895   59.128983\n",
       "20   20   63025520   321007      3280        0.005093   63.025520\n",
       "21   21   48129895   175650      2351        0.003649   48.129895\n",
       "22   22   51304566   224761      2206        0.004381   51.304566\n",
       "23    X  155270560   490861      5788        0.003161  155.270560\n",
       "24    Y   59373566    10375      3028        0.000175   59.373566\n",
       "25    *          0        0   1174990             NaN    0.000000"
      ]
     },
     "execution_count": 179,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idx_df[\"length_Mbs\"]=idx_df[\"length\"]/1e6\n",
    "idx_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa044bd6-ec1e-43ff-8916-4c00a66ba275",
   "metadata": {},
   "source": [
    "Subset the rows that have `reads_per_base` values greater than 0.005 and a chromosome length under 100 Mb."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "id": "d2efec08-e9e3-48cc-b27d-5a1f69861a58",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>length</th>\n",
       "      <th>mapped</th>\n",
       "      <th>unmapped</th>\n",
       "      <th>reads_per_base</th>\n",
       "      <th>length_Mbs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MT</td>\n",
       "      <td>16571</td>\n",
       "      <td>226</td>\n",
       "      <td>2</td>\n",
       "      <td>0.013638</td>\n",
       "      <td>0.016571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16</td>\n",
       "      <td>90354753</td>\n",
       "      <td>483073</td>\n",
       "      <td>5439</td>\n",
       "      <td>0.005346</td>\n",
       "      <td>90.354753</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>17</td>\n",
       "      <td>81195210</td>\n",
       "      <td>532624</td>\n",
       "      <td>5339</td>\n",
       "      <td>0.006560</td>\n",
       "      <td>81.195210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>19</td>\n",
       "      <td>59128983</td>\n",
       "      <td>407712</td>\n",
       "      <td>5882</td>\n",
       "      <td>0.006895</td>\n",
       "      <td>59.128983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>20</td>\n",
       "      <td>63025520</td>\n",
       "      <td>321007</td>\n",
       "      <td>3280</td>\n",
       "      <td>0.005093</td>\n",
       "      <td>63.025520</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   name    length  mapped  unmapped  reads_per_base  length_Mbs\n",
       "0    MT     16571     226         2        0.013638    0.016571\n",
       "16   16  90354753  483073      5439        0.005346   90.354753\n",
       "17   17  81195210  532624      5339        0.006560   81.195210\n",
       "19   19  59128983  407712      5882        0.006895   59.128983\n",
       "20   20  63025520  321007      3280        0.005093   63.025520"
      ]
     },
     "execution_count": 180,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idx_df[(idx_df[\"reads_per_base\"] > 0.005) & (idx_df[\"length_Mbs\"] < 100)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9451117a-8214-405e-a54f-87dffcbd6852",
   "metadata": {},
   "source": [
    "Delete the `reads_per_base` column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "db35acf7-08db-4c0b-8064-09f1406087d1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>length</th>\n",
       "      <th>mapped</th>\n",
       "      <th>unmapped</th>\n",
       "      <th>length_Mbs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MT</td>\n",
       "      <td>16571</td>\n",
       "      <td>226</td>\n",
       "      <td>2</td>\n",
       "      <td>0.016571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>249250621</td>\n",
       "      <td>1202908</td>\n",
       "      <td>11586</td>\n",
       "      <td>249.250621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>243199373</td>\n",
       "      <td>1179409</td>\n",
       "      <td>11950</td>\n",
       "      <td>243.199373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>198022430</td>\n",
       "      <td>956939</td>\n",
       "      <td>7894</td>\n",
       "      <td>198.022430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>191154276</td>\n",
       "      <td>841065</td>\n",
       "      <td>8496</td>\n",
       "      <td>191.154276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>180915260</td>\n",
       "      <td>839376</td>\n",
       "      <td>7313</td>\n",
       "      <td>180.915260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>171115067</td>\n",
       "      <td>851357</td>\n",
       "      <td>6930</td>\n",
       "      <td>171.115067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>159138663</td>\n",
       "      <td>639203</td>\n",
       "      <td>7728</td>\n",
       "      <td>159.138663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>146364022</td>\n",
       "      <td>951606</td>\n",
       "      <td>8874</td>\n",
       "      <td>146.364022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>141213431</td>\n",
       "      <td>603020</td>\n",
       "      <td>6937</td>\n",
       "      <td>141.213431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>135534747</td>\n",
       "      <td>692429</td>\n",
       "      <td>10325</td>\n",
       "      <td>135.534747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>135006516</td>\n",
       "      <td>660885</td>\n",
       "      <td>7635</td>\n",
       "      <td>135.006516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>133851895</td>\n",
       "      <td>728753</td>\n",
       "      <td>6993</td>\n",
       "      <td>133.851895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13</td>\n",
       "      <td>115169878</td>\n",
       "      <td>523741</td>\n",
       "      <td>4061</td>\n",
       "      <td>115.169878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>14</td>\n",
       "      <td>107349540</td>\n",
       "      <td>480831</td>\n",
       "      <td>4255</td>\n",
       "      <td>107.349540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>15</td>\n",
       "      <td>102531392</td>\n",
       "      <td>447954</td>\n",
       "      <td>3936</td>\n",
       "      <td>102.531392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16</td>\n",
       "      <td>90354753</td>\n",
       "      <td>483073</td>\n",
       "      <td>5439</td>\n",
       "      <td>90.354753</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>17</td>\n",
       "      <td>81195210</td>\n",
       "      <td>532624</td>\n",
       "      <td>5339</td>\n",
       "      <td>81.195210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>18</td>\n",
       "      <td>78077248</td>\n",
       "      <td>319957</td>\n",
       "      <td>4373</td>\n",
       "      <td>78.077248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>19</td>\n",
       "      <td>59128983</td>\n",
       "      <td>407712</td>\n",
       "      <td>5882</td>\n",
       "      <td>59.128983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>20</td>\n",
       "      <td>63025520</td>\n",
       "      <td>321007</td>\n",
       "      <td>3280</td>\n",
       "      <td>63.025520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>21</td>\n",
       "      <td>48129895</td>\n",
       "      <td>175650</td>\n",
       "      <td>2351</td>\n",
       "      <td>48.129895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>22</td>\n",
       "      <td>51304566</td>\n",
       "      <td>224761</td>\n",
       "      <td>2206</td>\n",
       "      <td>51.304566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>X</td>\n",
       "      <td>155270560</td>\n",
       "      <td>490861</td>\n",
       "      <td>5788</td>\n",
       "      <td>155.270560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Y</td>\n",
       "      <td>59373566</td>\n",
       "      <td>10375</td>\n",
       "      <td>3028</td>\n",
       "      <td>59.373566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1174990</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   name     length   mapped  unmapped  length_Mbs\n",
       "0    MT      16571      226         2    0.016571\n",
       "1     1  249250621  1202908     11586  249.250621\n",
       "2     2  243199373  1179409     11950  243.199373\n",
       "3     3  198022430   956939      7894  198.022430\n",
       "4     4  191154276   841065      8496  191.154276\n",
       "5     5  180915260   839376      7313  180.915260\n",
       "6     6  171115067   851357      6930  171.115067\n",
       "7     7  159138663   639203      7728  159.138663\n",
       "8     8  146364022   951606      8874  146.364022\n",
       "9     9  141213431   603020      6937  141.213431\n",
       "10   10  135534747   692429     10325  135.534747\n",
       "11   11  135006516   660885      7635  135.006516\n",
       "12   12  133851895   728753      6993  133.851895\n",
       "13   13  115169878   523741      4061  115.169878\n",
       "14   14  107349540   480831      4255  107.349540\n",
       "15   15  102531392   447954      3936  102.531392\n",
       "16   16   90354753   483073      5439   90.354753\n",
       "17   17   81195210   532624      5339   81.195210\n",
       "18   18   78077248   319957      4373   78.077248\n",
       "19   19   59128983   407712      5882   59.128983\n",
       "20   20   63025520   321007      3280   63.025520\n",
       "21   21   48129895   175650      2351   48.129895\n",
       "22   22   51304566   224761      2206   51.304566\n",
       "23    X  155270560   490861      5788  155.270560\n",
       "24    Y   59373566    10375      3028   59.373566\n",
       "25    *          0        0   1174990    0.000000"
      ]
     },
     "execution_count": 181,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idx_df = idx_df.drop(columns = [\"reads_per_base\"])\n",
    "idx_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d75d93b4-5729-466c-9ec4-913ad0a7ada3",
   "metadata": {},
   "source": [
    "# Exercise 3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de0af99c-1ae9-49a8-b975-2cacf50c8d39",
   "metadata": {
    "tags": []
   },
   "source": [
    "Read in GTF File \"Homo_sapiens_GRCh38_107_2.gtf” as a `DataFrame`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "id": "2733e4af-4126-4129-8fe0-95c633a7678b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_439204/2607183175.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(\"/project/clab0632/python/SHARED/python/2_python_data_science/data/Homo_sapiens_GRCh38_107_2.gtf\",sep='\\t',header=None)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1471887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   0               1            2        3        4  5  6  7  \\\n",
       "0  1  ensembl_havana         gene  1471765  1497848  .  +  .   \n",
       "1  1  ensembl_havana   transcript  1471765  1497848  .  +  .   \n",
       "2  1  ensembl_havana         exon  1471765  1472089  .  +  .   \n",
       "3  1  ensembl_havana          CDS  1471885  1472089  .  +  0   \n",
       "4  1  ensembl_havana  start_codon  1471885  1471887  .  +  0   \n",
       "\n",
       "                                                   8  \n",
       "0  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "1  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "2  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "3  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "4  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  "
      ]
     },
     "execution_count": 182,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\"/project/clab0632/python/SHARED/python/2_python_data_science/data/Homo_sapiens_GRCh38_107_2.gtf\",sep='\\t',header=None)\n",
    "df.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5ed1277-b4bb-4fd9-8c35-258ce18235db",
   "metadata": {},
   "source": [
    "Set the column names to: 'seqname', 'source', 'feature’, 'start', 'end', 'score', 'strand', 'frame’, and 'attribute'."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "id": "5f69a4fe-fd41-492a-8f30-aff5abe806a2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1471887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname          source      feature    start      end score strand frame  \\\n",
       "0       1  ensembl_havana         gene  1471765  1497848     .      +     .   \n",
       "1       1  ensembl_havana   transcript  1471765  1497848     .      +     .   \n",
       "2       1  ensembl_havana         exon  1471765  1472089     .      +     .   \n",
       "3       1  ensembl_havana          CDS  1471885  1472089     .      +     0   \n",
       "4       1  ensembl_havana  start_codon  1471885  1471887     .      +     0   \n",
       "\n",
       "                                           attribute  \n",
       "0  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "1  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "2  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "3  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "4  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  "
      ]
     },
     "execution_count": 183,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns=['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame','attribute']\n",
    "df.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19d2e1d9-f48d-4f4e-8a0b-d513aa0ec49f",
   "metadata": {
    "tags": []
   },
   "source": [
    "How many unique values are there for the \"source\" column?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "id": "5227afa4-9bb9-4a04-b256-724d8300bbbe",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "source\n",
       "havana                   2340968\n",
       "ensembl_havana            714608\n",
       "havana_tagene             170046\n",
       "ensembl                   139551\n",
       "mirbase                     5631\n",
       "ensembl_havana_tagene        275\n",
       "insdc                        144\n",
       "ensembl_tagene                21\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 211,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(df['source'].nunique())\n",
    "df['source'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "828bac8e-f1cc-4639-8ee6-cb126119e123",
   "metadata": {},
   "source": [
    "Subset all rows with the “ensembl_havana” source and assign to new DataFrame called `ens_df`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "id": "b599cd77-e151-42a3-9612-a1d799a31020",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1471887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname          source      feature    start      end score strand frame  \\\n",
       "0       1  ensembl_havana         gene  1471765  1497848     .      +     .   \n",
       "1       1  ensembl_havana   transcript  1471765  1497848     .      +     .   \n",
       "2       1  ensembl_havana         exon  1471765  1472089     .      +     .   \n",
       "3       1  ensembl_havana          CDS  1471885  1472089     .      +     0   \n",
       "4       1  ensembl_havana  start_codon  1471885  1471887     .      +     0   \n",
       "\n",
       "                                           attribute  \n",
       "0  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "1  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "2  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "3  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "4  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  "
      ]
     },
     "execution_count": 185,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ens_df = df[df[\"source\"] == \"ensembl_havana\"]\n",
    "ens_df.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b30adf7-3c0a-4025-adc0-f29660339e29",
   "metadata": {},
   "source": [
    "What is the shape of `ens_df`?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "id": "9a2ca8b2-b231-4eac-b3ee-7de7f3ac8858",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(714608, 9)"
      ]
     },
     "execution_count": 186,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ens_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71f6dfbd-9350-4d11-a3cb-6500db112658",
   "metadata": {},
   "source": [
    "Group `ens_df` by feature column and apply the `value_counts` method to get the number of rows in each feature."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "id": "defd5520-dfc3-450d-8e72-107464a1abf4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "feature\n",
       "CDS                262659\n",
       "Selenocysteine         13\n",
       "exon               285809\n",
       "five_prime_utr      38184\n",
       "gene                19285\n",
       "start_codon         25889\n",
       "stop_codon          25906\n",
       "three_prime_utr     30118\n",
       "transcript          26745\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 187,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ens_df.groupby(\"feature\", observed = False)['feature'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53355ecf-974e-457c-a060-180405d556ec",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Exercise 4"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52d16dd4-df24-4923-873c-1d8850e3f62f",
   "metadata": {},
   "source": [
    "Create a new DataFrame called `genes` by subsetting all rows with “gene” feature type in `ens_df`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "id": "9f042135-0567-493e-915d-7c15f97f6612",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3069168</td>\n",
       "      <td>3438621</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000142611\"; gene_version \"17\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2403964</td>\n",
       "      <td>2413797</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000157911\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>517</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2228319</td>\n",
       "      <td>2310213</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000157933\"; gene_version \"10\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>580</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>10472288</td>\n",
       "      <td>10630758</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000142655\"; gene_version \"13\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3365880</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>6318378</td>\n",
       "      <td>6360415</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000278903\"; gene_version \"5\"; g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3365953</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>33915534</td>\n",
       "      <td>33977703</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237945\"; gene_version \"9\"; g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3367972</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>9068325</td>\n",
       "      <td>9129773</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000188681\"; gene_version \"13\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3368345</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>14236206</td>\n",
       "      <td>14362754</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000243064\"; gene_version \"10\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3369764</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>46324124</td>\n",
       "      <td>46445769</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160299\"; gene_version \"19\"; ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19285 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname          source feature     start       end score strand  \\\n",
       "0             1  ensembl_havana    gene   1471765   1497848     .      +   \n",
       "132           1  ensembl_havana    gene   3069168   3438621     .      +   \n",
       "361           1  ensembl_havana    gene   2403964   2413797     .      -   \n",
       "517           1  ensembl_havana    gene   2228319   2310213     .      +   \n",
       "580           1  ensembl_havana    gene  10472288  10630758     .      +   \n",
       "...         ...             ...     ...       ...       ...   ...    ...   \n",
       "3365880      21  ensembl_havana    gene   6318378   6360415     .      +   \n",
       "3365953      21  ensembl_havana    gene  33915534  33977703     .      +   \n",
       "3367972      21  ensembl_havana    gene   9068325   9129773     .      -   \n",
       "3368345      21  ensembl_havana    gene  14236206  14362754     .      +   \n",
       "3369764      21  ensembl_havana    gene  46324124  46445769     .      +   \n",
       "\n",
       "        frame                                          attribute  \n",
       "0           .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "132         .  gene_id \"ENSG00000142611\"; gene_version \"17\"; ...  \n",
       "361         .  gene_id \"ENSG00000157911\"; gene_version \"11\"; ...  \n",
       "517         .  gene_id \"ENSG00000157933\"; gene_version \"10\"; ...  \n",
       "580         .  gene_id \"ENSG00000142655\"; gene_version \"13\"; ...  \n",
       "...       ...                                                ...  \n",
       "3365880     .  gene_id \"ENSG00000278903\"; gene_version \"5\"; g...  \n",
       "3365953     .  gene_id \"ENSG00000237945\"; gene_version \"9\"; g...  \n",
       "3367972     .  gene_id \"ENSG00000188681\"; gene_version \"13\"; ...  \n",
       "3368345     .  gene_id \"ENSG00000243064\"; gene_version \"10\"; ...  \n",
       "3369764     .  gene_id \"ENSG00000160299\"; gene_version \"19\"; ...  \n",
       "\n",
       "[19285 rows x 9 columns]"
      ]
     },
     "execution_count": 265,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = ens_df[ens_df['feature']=='gene']\n",
    "genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef5a0aa4-35ce-4a60-ad06-d269148e17a7",
   "metadata": {},
   "source": [
    "Create a new column called “gene_length” that contains the length of each gene in dataframe genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 271,
   "id": "0a15d22f-2500-4fce-8206-c757c8bde6e7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2913291</th>\n",
       "      <td>17</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>43131208</td>\n",
       "      <td>43131270</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000285057\"; gene_version \"1\"; g...</td>\n",
       "      <td>63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2118533</th>\n",
       "      <td>12</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>93083598</td>\n",
       "      <td>93083675</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000256393\"; gene_version \"1\"; g...</td>\n",
       "      <td>78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2362811</th>\n",
       "      <td>14</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>22852616</td>\n",
       "      <td>22852728</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000212335\"; gene_version \"1\"; g...</td>\n",
       "      <td>113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>904441</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>90838903</td>\n",
       "      <td>90839037</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000187653\"; gene_version \"11\"; ...</td>\n",
       "      <td>135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1133423</th>\n",
       "      <td>6</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>73373108</td>\n",
       "      <td>73373263</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000235174\"; gene_version \"1\"; g...</td>\n",
       "      <td>156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129959</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>27649419</td>\n",
       "      <td>27649610</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000235912\"; gene_version \"1\"; g...</td>\n",
       "      <td>192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1847776</th>\n",
       "      <td>11</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>82689559</td>\n",
       "      <td>82689768</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227097\"; gene_version \"5\"; g...</td>\n",
       "      <td>210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2938518</th>\n",
       "      <td>18</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>52409042</td>\n",
       "      <td>52409257</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000280098\"; gene_version \"1\"; g...</td>\n",
       "      <td>216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>675285</th>\n",
       "      <td>3</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>186596976</td>\n",
       "      <td>186597200</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000283390\"; gene_version \"1\"; g...</td>\n",
       "      <td>225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>988497</th>\n",
       "      <td>5</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>62512246</td>\n",
       "      <td>62512482</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000268942\"; gene_version \"2\"; g...</td>\n",
       "      <td>237</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname          source feature      start        end score strand  \\\n",
       "2913291      17  ensembl_havana    gene   43131208   43131270     .      +   \n",
       "2118533      12  ensembl_havana    gene   93083598   93083675     .      -   \n",
       "2362811      14  ensembl_havana    gene   22852616   22852728     .      -   \n",
       "904441        4  ensembl_havana    gene   90838903   90839037     .      -   \n",
       "1133423       6  ensembl_havana    gene   73373108   73373263     .      -   \n",
       "129959        1  ensembl_havana    gene   27649419   27649610     .      +   \n",
       "1847776      11  ensembl_havana    gene   82689559   82689768     .      +   \n",
       "2938518      18  ensembl_havana    gene   52409042   52409257     .      +   \n",
       "675285        3  ensembl_havana    gene  186596976  186597200     .      +   \n",
       "988497        5  ensembl_havana    gene   62512246   62512482     .      -   \n",
       "\n",
       "        frame                                          attribute  gene_length  \n",
       "2913291     .  gene_id \"ENSG00000285057\"; gene_version \"1\"; g...           63  \n",
       "2118533     .  gene_id \"ENSG00000256393\"; gene_version \"1\"; g...           78  \n",
       "2362811     .  gene_id \"ENSG00000212335\"; gene_version \"1\"; g...          113  \n",
       "904441      .  gene_id \"ENSG00000187653\"; gene_version \"11\"; ...          135  \n",
       "1133423     .  gene_id \"ENSG00000235174\"; gene_version \"1\"; g...          156  \n",
       "129959      .  gene_id \"ENSG00000235912\"; gene_version \"1\"; g...          192  \n",
       "1847776     .  gene_id \"ENSG00000227097\"; gene_version \"5\"; g...          210  \n",
       "2938518     .  gene_id \"ENSG00000280098\"; gene_version \"1\"; g...          216  \n",
       "675285      .  gene_id \"ENSG00000283390\"; gene_version \"1\"; g...          225  \n",
       "988497      .  gene_id \"ENSG00000268942\"; gene_version \"2\"; g...          237  "
      ]
     },
     "execution_count": 271,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = genes.assign(gene_length = genes[\"end\"]-genes[\"start\"]+1)\n",
    "genes.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "156894f8-e87b-4690-8de3-d0b3dcd82513",
   "metadata": {},
   "source": [
    "Print the mean and median of the gene lengths."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 272,
   "id": "f3f3b383-3618-4e7b-b01f-56e6e5702c25",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mean gene length:67672.02224526834\n",
      "Median gene length:27365.0\n"
     ]
    }
   ],
   "source": [
    "print(f'Mean gene length:{genes[\"gene_length\"].mean()}')\n",
    "print(f'Median gene length:{genes[\"gene_length\"].median()}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1c311d3-7a5c-4f98-b758-221c55514ebb",
   "metadata": {},
   "source": [
    "Find the top 10 longest genes and print their gene identifier."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "id": "f8599ab3-46b8-46c9-b834-34e1cf03c260",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2679474    ENSG00000078328\n",
       "1297140    ENSG00000174469\n",
       "1733179    ENSG00000153707\n",
       "1514121    ENSG00000198947\n",
       "1850816    ENSG00000150672\n",
       "1601567    ENSG00000183117\n",
       "3030998    ENSG00000172264\n",
       "1104549    ENSG00000188107\n",
       "2041854    ENSG00000183230\n",
       "2040306    ENSG00000150275\n",
       "Name: attribute, dtype: object"
      ]
     },
     "execution_count": 268,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = genes.sort_values(by=[\"gene_length\"], ascending =False)\n",
    "top10=genes.head(10)\n",
    "top10['attribute'].str[9:24]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 278,
   "id": "82ca9832-2846-4a48-8607-3c134cfc2b25",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2679474</th>\n",
       "      <td>ENSG00000078328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1297140</th>\n",
       "      <td>ENSG00000174469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1733179</th>\n",
       "      <td>ENSG00000153707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1514121</th>\n",
       "      <td>ENSG00000198947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1850816</th>\n",
       "      <td>ENSG00000150672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1601567</th>\n",
       "      <td>ENSG00000183117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3030998</th>\n",
       "      <td>ENSG00000172264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1104549</th>\n",
       "      <td>ENSG00000188107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2041854</th>\n",
       "      <td>ENSG00000183230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2040306</th>\n",
       "      <td>ENSG00000150275</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       0\n",
       "2679474  ENSG00000078328\n",
       "1297140  ENSG00000174469\n",
       "1733179  ENSG00000153707\n",
       "1514121  ENSG00000198947\n",
       "1850816  ENSG00000150672\n",
       "1601567  ENSG00000183117\n",
       "3030998  ENSG00000172264\n",
       "1104549  ENSG00000188107\n",
       "2041854  ENSG00000183230\n",
       "2040306  ENSG00000150275"
      ]
     },
     "execution_count": 278,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes.nlargest(\n",
    "    n = 10,\n",
    "    columns = \"gene_length\"\n",
    ")[\"attribute\"].str.extract(\n",
    "    pat = r'.*gene_id \"([A-Za-z0-9]+)\".*')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8676700d-4fdf-4633-abde-c39a527701ca",
   "metadata": {},
   "source": [
    "Find the lowest 20 shortest genes and print their gene identifier."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 270,
   "id": "85cffbfd-7c9e-47bf-9746-228348579da7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2913291    ENSG00000285057\n",
       "2118533    ENSG00000256393\n",
       "2362811    ENSG00000212335\n",
       "904441     ENSG00000187653\n",
       "1133423    ENSG00000235174\n",
       "129959     ENSG00000235912\n",
       "1847776    ENSG00000227097\n",
       "2938518    ENSG00000280098\n",
       "675285     ENSG00000283390\n",
       "988497     ENSG00000268942\n",
       "1967063    ENSG00000225976\n",
       "3255568    ENSG00000185275\n",
       "3363759    ENSG00000231068\n",
       "2274799    ENSG00000227081\n",
       "1268412    ENSG00000238072\n",
       "105987     ENSG00000185962\n",
       "2594158    ENSG00000230330\n",
       "2526556    ENSG00000234664\n",
       "3363751    ENSG00000206104\n",
       "3363727    ENSG00000244624\n",
       "Name: attribute, dtype: object"
      ]
     },
     "execution_count": 270,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = genes.sort_values(by=[\"gene_length\"])\n",
    "low20 = genes.head(20)\n",
    "low20['attribute'].str[9:24]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe7eeed1-71d5-43ca-8edb-01048012ad80",
   "metadata": {
    "tags": []
   },
   "source": [
    "Split the “attribute” column of the dataframe `genes` using `;` as delimiter, store the result in a new dataframe called `attributes`, then concatenante `genes` and `attributes`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "id": "49bca6d8-84f4-4a93-9a20-c9025c37a078",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_version</th>\n",
       "      <th>gene_source</th>\n",
       "      <th>gene_biotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2679474</th>\n",
       "      <td>ENSG00000078328</td>\n",
       "      <td>23</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1297140</th>\n",
       "      <td>ENSG00000174469</td>\n",
       "      <td>23</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1733179</th>\n",
       "      <td>ENSG00000153707</td>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1514121</th>\n",
       "      <td>ENSG00000198947</td>\n",
       "      <td>18</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1850816</th>\n",
       "      <td>ENSG00000150672</td>\n",
       "      <td>18</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1133423</th>\n",
       "      <td>ENSG00000235174</td>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>904441</th>\n",
       "      <td>ENSG00000187653</td>\n",
       "      <td>11</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2362811</th>\n",
       "      <td>ENSG00000212335</td>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>misc_RNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2118533</th>\n",
       "      <td>ENSG00000256393</td>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2913291</th>\n",
       "      <td>ENSG00000285057</td>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19285 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 gene_id gene_version     gene_source  \\\n",
       "2679474  ENSG00000078328           23  ensembl_havana   \n",
       "1297140  ENSG00000174469           23  ensembl_havana   \n",
       "1733179  ENSG00000153707           19  ensembl_havana   \n",
       "1514121  ENSG00000198947           18  ensembl_havana   \n",
       "1850816  ENSG00000150672           18  ensembl_havana   \n",
       "...                  ...          ...             ...   \n",
       "1133423  ENSG00000235174            1  ensembl_havana   \n",
       "904441   ENSG00000187653           11  ensembl_havana   \n",
       "2362811  ENSG00000212335            1  ensembl_havana   \n",
       "2118533  ENSG00000256393            1  ensembl_havana   \n",
       "2913291  ENSG00000285057            1  ensembl_havana   \n",
       "\n",
       "                               gene_biotype  \n",
       "2679474                      protein_coding  \n",
       "1297140                      protein_coding  \n",
       "1733179                      protein_coding  \n",
       "1514121                      protein_coding  \n",
       "1850816                      protein_coding  \n",
       "...                                     ...  \n",
       "1133423                processed_pseudogene  \n",
       "904441                 processed_pseudogene  \n",
       "2362811                            misc_RNA  \n",
       "2118533                processed_pseudogene  \n",
       "2913291  transcribed_unprocessed_pseudogene  \n",
       "\n",
       "[19285 rows x 4 columns]"
      ]
     },
     "execution_count": 277,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attributes = pd.DataFrame(genes['attribute'].str.split(\";\",expand=True))\n",
    "attributes.columns=['gene_id','gene_version','gene_source','gene_biotype']\n",
    "attributes['gene_id']=attributes['gene_id'].str[9:-1]\n",
    "attributes['gene_version']=attributes['gene_version'].str[15:-1]\n",
    "attributes['gene_source']=attributes['gene_source'].str[14:-1]\n",
    "attributes['gene_biotype']=attributes['gene_biotype'].str[15:-1]\n",
    "attributes\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "id": "2679400d-c1af-4c62-ad1d-eda788432217",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_length</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_version</th>\n",
       "      <th>gene_source</th>\n",
       "      <th>gene_biotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>17</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>43131208</td>\n",
       "      <td>43131270</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000285057\"; gene_version \"1\"; g...</td>\n",
       "      <td>63</td>\n",
       "      <td>ENSG00000285057</td>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>93083598</td>\n",
       "      <td>93083675</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000256393\"; gene_version \"1\"; g...</td>\n",
       "      <td>78</td>\n",
       "      <td>ENSG00000256393</td>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>14</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>22852616</td>\n",
       "      <td>22852728</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000212335\"; gene_version \"1\"; g...</td>\n",
       "      <td>113</td>\n",
       "      <td>ENSG00000212335</td>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>misc_RNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>90838903</td>\n",
       "      <td>90839037</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000187653\"; gene_version \"11\"; ...</td>\n",
       "      <td>135</td>\n",
       "      <td>ENSG00000187653</td>\n",
       "      <td>11</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>73373108</td>\n",
       "      <td>73373263</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000235174\"; gene_version \"1\"; g...</td>\n",
       "      <td>156</td>\n",
       "      <td>ENSG00000235174</td>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19280</th>\n",
       "      <td>11</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>83455012</td>\n",
       "      <td>85627922</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000150672\"; gene_version \"18\"; ...</td>\n",
       "      <td>2172911</td>\n",
       "      <td>ENSG00000150672</td>\n",
       "      <td>18</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19281</th>\n",
       "      <td>X</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>31097677</td>\n",
       "      <td>33339609</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000198947\"; gene_version \"18\"; ...</td>\n",
       "      <td>2241933</td>\n",
       "      <td>ENSG00000198947</td>\n",
       "      <td>18</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19282</th>\n",
       "      <td>9</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>8314246</td>\n",
       "      <td>10613002</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000153707\"; gene_version \"19\"; ...</td>\n",
       "      <td>2298757</td>\n",
       "      <td>ENSG00000153707</td>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19283</th>\n",
       "      <td>7</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>146116002</td>\n",
       "      <td>148420998</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000174469\"; gene_version \"23\"; ...</td>\n",
       "      <td>2304997</td>\n",
       "      <td>ENSG00000174469</td>\n",
       "      <td>23</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19284</th>\n",
       "      <td>16</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>5239802</td>\n",
       "      <td>7713340</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000078328\"; gene_version \"23\"; ...</td>\n",
       "      <td>2473539</td>\n",
       "      <td>ENSG00000078328</td>\n",
       "      <td>23</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19285 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqname          source feature      start        end score strand  \\\n",
       "0          17  ensembl_havana    gene   43131208   43131270     .      +   \n",
       "1          12  ensembl_havana    gene   93083598   93083675     .      -   \n",
       "2          14  ensembl_havana    gene   22852616   22852728     .      -   \n",
       "3           4  ensembl_havana    gene   90838903   90839037     .      -   \n",
       "4           6  ensembl_havana    gene   73373108   73373263     .      -   \n",
       "...       ...             ...     ...        ...        ...   ...    ...   \n",
       "19280      11  ensembl_havana    gene   83455012   85627922     .      -   \n",
       "19281       X  ensembl_havana    gene   31097677   33339609     .      -   \n",
       "19282       9  ensembl_havana    gene    8314246   10613002     .      -   \n",
       "19283       7  ensembl_havana    gene  146116002  148420998     .      +   \n",
       "19284      16  ensembl_havana    gene    5239802    7713340     .      +   \n",
       "\n",
       "      frame                                          attribute  gene_length  \\\n",
       "0         .  gene_id \"ENSG00000285057\"; gene_version \"1\"; g...           63   \n",
       "1         .  gene_id \"ENSG00000256393\"; gene_version \"1\"; g...           78   \n",
       "2         .  gene_id \"ENSG00000212335\"; gene_version \"1\"; g...          113   \n",
       "3         .  gene_id \"ENSG00000187653\"; gene_version \"11\"; ...          135   \n",
       "4         .  gene_id \"ENSG00000235174\"; gene_version \"1\"; g...          156   \n",
       "...     ...                                                ...          ...   \n",
       "19280     .  gene_id \"ENSG00000150672\"; gene_version \"18\"; ...      2172911   \n",
       "19281     .  gene_id \"ENSG00000198947\"; gene_version \"18\"; ...      2241933   \n",
       "19282     .  gene_id \"ENSG00000153707\"; gene_version \"19\"; ...      2298757   \n",
       "19283     .  gene_id \"ENSG00000174469\"; gene_version \"23\"; ...      2304997   \n",
       "19284     .  gene_id \"ENSG00000078328\"; gene_version \"23\"; ...      2473539   \n",
       "\n",
       "               gene_id gene_version     gene_source  \\\n",
       "0      ENSG00000285057            1  ensembl_havana   \n",
       "1      ENSG00000256393            1  ensembl_havana   \n",
       "2      ENSG00000212335            1  ensembl_havana   \n",
       "3      ENSG00000187653           11  ensembl_havana   \n",
       "4      ENSG00000235174            1  ensembl_havana   \n",
       "...                ...          ...             ...   \n",
       "19280  ENSG00000150672           18  ensembl_havana   \n",
       "19281  ENSG00000198947           18  ensembl_havana   \n",
       "19282  ENSG00000153707           19  ensembl_havana   \n",
       "19283  ENSG00000174469           23  ensembl_havana   \n",
       "19284  ENSG00000078328           23  ensembl_havana   \n",
       "\n",
       "                             gene_biotype  \n",
       "0      transcribed_unprocessed_pseudogene  \n",
       "1                    processed_pseudogene  \n",
       "2                                misc_RNA  \n",
       "3                    processed_pseudogene  \n",
       "4                    processed_pseudogene  \n",
       "...                                   ...  \n",
       "19280                      protein_coding  \n",
       "19281                      protein_coding  \n",
       "19282                      protein_coding  \n",
       "19283                      protein_coding  \n",
       "19284                      protein_coding  \n",
       "\n",
       "[19285 rows x 14 columns]"
      ]
     },
     "execution_count": 245,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.concat([genes, attributes], axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ef23030-d210-4f64-bd97-df7420c0d6fa",
   "metadata": {
    "tags": []
   },
   "source": [
    "Using `genes`, count the number of genes per chromosome and add the number of genes per chomosome into the `idx_df` from part 1 using the `merge` method. NOTE: use `astype()` and `pd.StringDtype()` to convert chromosome names to `string` type."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "id": "ba02fdd5-b295-4e24-b5af-ab8f19a68675",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>length</th>\n",
       "      <th>mapped</th>\n",
       "      <th>unmapped</th>\n",
       "      <th>gene_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MT</td>\n",
       "      <td>16571</td>\n",
       "      <td>226</td>\n",
       "      <td>2</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>249250621</td>\n",
       "      <td>1202908</td>\n",
       "      <td>11586</td>\n",
       "      <td>2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>243199373</td>\n",
       "      <td>1179409</td>\n",
       "      <td>11950</td>\n",
       "      <td>1222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>198022430</td>\n",
       "      <td>956939</td>\n",
       "      <td>7894</td>\n",
       "      <td>1040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>191154276</td>\n",
       "      <td>841065</td>\n",
       "      <td>8496</td>\n",
       "      <td>736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>180915260</td>\n",
       "      <td>839376</td>\n",
       "      <td>7313</td>\n",
       "      <td>835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>171115067</td>\n",
       "      <td>851357</td>\n",
       "      <td>6930</td>\n",
       "      <td>995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>159138663</td>\n",
       "      <td>639203</td>\n",
       "      <td>7728</td>\n",
       "      <td>875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>146364022</td>\n",
       "      <td>951606</td>\n",
       "      <td>8874</td>\n",
       "      <td>660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>141213431</td>\n",
       "      <td>603020</td>\n",
       "      <td>6937</td>\n",
       "      <td>752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>135534747</td>\n",
       "      <td>692429</td>\n",
       "      <td>10325</td>\n",
       "      <td>725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>135006516</td>\n",
       "      <td>660885</td>\n",
       "      <td>7635</td>\n",
       "      <td>1261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>133851895</td>\n",
       "      <td>728753</td>\n",
       "      <td>6993</td>\n",
       "      <td>998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13</td>\n",
       "      <td>115169878</td>\n",
       "      <td>523741</td>\n",
       "      <td>4061</td>\n",
       "      <td>319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>14</td>\n",
       "      <td>107349540</td>\n",
       "      <td>480831</td>\n",
       "      <td>4255</td>\n",
       "      <td>616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>15</td>\n",
       "      <td>102531392</td>\n",
       "      <td>447954</td>\n",
       "      <td>3936</td>\n",
       "      <td>577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16</td>\n",
       "      <td>90354753</td>\n",
       "      <td>483073</td>\n",
       "      <td>5439</td>\n",
       "      <td>823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>17</td>\n",
       "      <td>81195210</td>\n",
       "      <td>532624</td>\n",
       "      <td>5339</td>\n",
       "      <td>1138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>18</td>\n",
       "      <td>78077248</td>\n",
       "      <td>319957</td>\n",
       "      <td>4373</td>\n",
       "      <td>265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>19</td>\n",
       "      <td>59128983</td>\n",
       "      <td>407712</td>\n",
       "      <td>5882</td>\n",
       "      <td>1362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>20</td>\n",
       "      <td>63025520</td>\n",
       "      <td>321007</td>\n",
       "      <td>3280</td>\n",
       "      <td>536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>21</td>\n",
       "      <td>48129895</td>\n",
       "      <td>175650</td>\n",
       "      <td>2351</td>\n",
       "      <td>217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>22</td>\n",
       "      <td>51304566</td>\n",
       "      <td>224761</td>\n",
       "      <td>2206</td>\n",
       "      <td>452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>X</td>\n",
       "      <td>155270560</td>\n",
       "      <td>490861</td>\n",
       "      <td>5788</td>\n",
       "      <td>811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Y</td>\n",
       "      <td>59373566</td>\n",
       "      <td>10375</td>\n",
       "      <td>3028</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1174990</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   name     length   mapped  unmapped  gene_count\n",
       "0    MT      16571      226         2        <NA>\n",
       "1     1  249250621  1202908     11586        2009\n",
       "2     2  243199373  1179409     11950        1222\n",
       "3     3  198022430   956939      7894        1040\n",
       "4     4  191154276   841065      8496         736\n",
       "5     5  180915260   839376      7313         835\n",
       "6     6  171115067   851357      6930         995\n",
       "7     7  159138663   639203      7728         875\n",
       "8     8  146364022   951606      8874         660\n",
       "9     9  141213431   603020      6937         752\n",
       "10   10  135534747   692429     10325         725\n",
       "11   11  135006516   660885      7635        1261\n",
       "12   12  133851895   728753      6993         998\n",
       "13   13  115169878   523741      4061         319\n",
       "14   14  107349540   480831      4255         616\n",
       "15   15  102531392   447954      3936         577\n",
       "16   16   90354753   483073      5439         823\n",
       "17   17   81195210   532624      5339        1138\n",
       "18   18   78077248   319957      4373         265\n",
       "19   19   59128983   407712      5882        1362\n",
       "20   20   63025520   321007      3280         536\n",
       "21   21   48129895   175650      2351         217\n",
       "22   22   51304566   224761      2206         452\n",
       "23    X  155270560   490861      5788         811\n",
       "24    Y   59373566    10375      3028          61\n",
       "25    *          0        0   1174990        <NA>"
      ]
     },
     "execution_count": 229,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Convert chromosome names to string type in both data sets\n",
    "idx_df = idx_df.astype({'name': pd.StringDtype()})\n",
    "genes = genes.astype({'seqname': pd.StringDtype()})\n",
    "# Count the number of genes per chromosome\n",
    "gene_per_chr=pd.DataFrame(genes['seqname'].value_counts().reset_index())\n",
    "gene_per_chr.columns = ['name', 'gene_count']\n",
    "gene_per_chr = gene_per_chr.astype({'name': pd.StringDtype()})\n",
    "gene_per_chr\n",
    "# Add that information to idx_df\n",
    "idx_df1=idx_df.merge(gene_per_chr, on=\"name\", how=\"outer\")\n",
    "idx_df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0cf00ff3-5048-4290-8136-e88581f8f6c8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
